### ISSUE TITLE
WIP: Void built with gcc-6.2.1

### ISSUE THREAD

**Issue by pullmoll** [voidlinux/void-packages#4825]
#### I started over to build Void Linux with `gcc-6.2.1`

The hint came from @chneukirchen that  [Arch Linux](https://www.archlinux.org/packages/core/x86_64/gcc/) uses gcc-6.2.1.

The compiler was cloned with

```
git clone git://gcc.gnu.org/git/gcc.git gcc-6.2.1
cd git-6.2.1
git checkout gcc-6-branch
cd ..
```

Then a tarball was created with

```
find gcc-6.2.1 \! -type d -print0 | \
    sort -z | \
    grep -z -v \.git/ | \
    tar cvf - -T - | \
    xz -v -9 -T 0 > gcc-6.2.1.tar.xz
```

I created an `etc/conf` with the following contents

```
XBPS_MASTERDIR=/tmp/gcc6
XBPS_HOSTDIR=/home/void/void-gcc6
XBPS_CHROOT_CMD=uunshare
XBPS_MAKEJOBS=12
XBPS_CCACHE=yes
XBPS_DISTFILES_MIRROR="https://sources.voidlinux.eu/ http://mirror.netcologne.de/gentoo/distfiles"
```

and then did run `./xbp-src zap` and `./xbps-src bootstrap`.
#### The gcc-6.2.1 homegrown tarball of 2016-11-11 is available at

```
distfiles="http://distfiles.voidlinux.de/gcc-6.2.1/gcc-6.2.1.tar.xz"
checksum=6bc5af0d2138bb8492c62ced00776b76bbc98b7297287aaa55adbe7d07786ac8
```
#### The repository including the patches required to bootstrap with gcc-6.2.1 is available at

https://github.com/pullmoll/void-packages-gcc6.2.1
#### Packages built will be uploaded to

| Everything built | Fast, but only a subset of pkgs [<a href="#space">1</a>] |
| --- | --- |
| http://voidlinux.myds.me/repo/current/ | http://repo.voidlinux.de/current |

**Note:** The musl libc pkgs are all in the same directory so as to save space for noarch pkgs.
#### Live ISOs created from these pkgs will be uploaded to

http://voidlinux.myds.me/repo/live/
#### Current package stats - updated  every now and then

| Architecture | Package count |
| --- | --- |
| x86_64 | 7028 |
| x86_64-musl | 6877 |
| i686 | 7043 |
#### Issues: open and solved
- [ ] **android-tools**: See [build log](http://misc.voidlinux.de/failed/android-tools__do_build.log) `char16_t` may be gcc specific, so perhaps `CXXFLAGS="-std=gnu++11"` or similar.
- [ ] **armagetronad**: See [build log](http://misc.voidlinux.de/failed/armagetronad__do_build.log) `tOutput::operator<<(tOutput const&)` isn't exported. `-fvisibility=...` issue?
- [ ] **audacity**: See [build log](http://misc.voidlinux.de/failed/audacity__do_build.log) several `return false;` where a `return NULL;` or `return nullptr;` is right. I think upstream fixed this. Add a patch?
- [ ] **avogadro**: See [build log](http://misc.voidlinux.de/failed/avogadro__do_build.log) replacing `signbit(m)` with `signbit((float)m)` might fix this.
- [ ] **avr-gcc**: See [build log](http://misc.voidlinux.de/failed/avr-gcc__do_build.log) Perhaps gcc-5.3.0 here can be updated to 6.2.1 as well.
- [ ] **beignet**: See [build log](http://misc.voidlinux.de/failed/beignet__do_build.log)
- [ ] **bwm-ng**: See [build log](http://misc.voidlinux.de/failed/bwm-ng__do_build.log)
- [ ] **calf**: See [build log](http://misc.voidlinux.de/failed/calf__do_build.log)
- [ ] **chromium-bsu**: See [build log](http://misc.voidlinux.de/failed/chromium-bsu__do_build.log) Patch to use `fabs(3)` instead of `abs(3)`.
- [ ] **clamav**: See [build log](http://misc.voidlinux.de/failed/clamav__do_build.log)
- [ ] **codeblocks**: See [build log](http://misc.voidlinux.de/failed/codeblocks__do_build.log)
- [ ] **crda**: See [build log](http://misc.voidlinux.de/failed/crda__do_build.log)
- [ ] **curlpp**: See [build log](http://misc.voidlinux.de/failed/curlpp__do_build.log) Add `CXXFLAGS="-Wno-error=deprecated-declarations"` to the template.
- [ ] **darkice**: See [build log](http://misc.voidlinux.de/failed/darkice__do_build.log)
- [ ] **desmume**: See [build log](http://misc.voidlinux.de/failed/desmume__do_build.log)
- [ ] **docker**: See [build log](http://misc.voidlinux.de/failed/docker__do_build.log)
- [ ] **dunelegacy**: See [build log](http://misc.voidlinux.de/failed/dunelegacy__do_build.log)
- [ ] **dunst**: See [build log](http://misc.voidlinux.de/failed/dunst__do_build.log)
- [ ] **enigma**: See [build log](http://misc.voidlinux.de/failed/enigma__do_build.log)
- [ ] **foobillard++**: See [build log](http://misc.voidlinux.de/failed/foobillard++__do_build.log)
- [ ] **fs-uae**: See [build log](http://misc.voidlinux.de/failed/fs-uae__do_build.log)
- [ ] **gdash**: See [build log](http://misc.voidlinux.de/failed/gdash__do_build.log)
- [ ] **gnome-shell-mousewheel-zoom**: See [build log](http://misc.voidlinux.de/failed/gnome-shell-mousewheel-zoom__do_build.log)
- [ ] **gpsd**: See [build log](http://misc.voidlinux.de/failed/gpsd__do_build.log)
- [ ] **gummiboot**: See [build log](http://misc.voidlinux.de/failed/gummiboot__do_build.log)
- [ ] **hardinfo**: See [build log](http://misc.voidlinux.de/failed/hardinfo__do_build.log)
- [ ] **ifenslave**: See [build log](http://misc.voidlinux.de/failed/ifenslave__do_build.log)
- [ ] **ifstatus**: See [build log](http://misc.voidlinux.de/failed/ifstatus__do_build.log)
- [ ] **iperf**: See [build log](http://misc.voidlinux.de/failed/iperf__do_build.log)
- [ ] **iscan**: See [build log](http://misc.voidlinux.de/failed/iscan__do_build.log)
- [ ] **ispc**: See [build log](http://misc.voidlinux.de/failed/ispc__do_build.log)
- [ ] **john**: See [build log](http://misc.voidlinux.de/failed/john__do_build.log)
- [ ] **js**: See [build log](http://misc.voidlinux.de/failed/js__do_build.log)
- [ ] **ksh**: See [build log](http://misc.voidlinux.de/failed/ksh__do_build.log)
- [ ] **ktorrent**: See [build log](http://misc.voidlinux.de/failed/ktorrent__do_build.log)
- [ ] **kvirc**: See [build log](http://misc.voidlinux.de/failed/kvirc__do_build.log)
- [ ] **ldc**: See [build log](http://misc.voidlinux.de/failed/ldc__do_build.log)
- [ ] **ledger**: See [build log](http://misc.voidlinux.de/failed/ledger__do_build.log)
- [ ] **libabw**: See [build log](http://misc.voidlinux.de/failed/libabw__do_build.log)
- [ ] **libe-book**: See [build log](http://misc.voidlinux.de/failed/libe-book__do_build.log)
- [ ] **libewf**: See [build log](http://misc.voidlinux.de/failed/libewf__do_build.log)
- [ ] **lightdm-gtk-greeter**: See [build log](http://misc.voidlinux.de/failed/lightdm-gtk-greeter__do_build.log)
- [ ] **lightdm-webkit-greeter**: See [build log](http://misc.voidlinux.de/failed/lightdm-webkit-greeter__do_build.log)
- [ ] **linux3.18**: See [build log](http://misc.voidlinux.de/failed/linux3.18__do_build.log)
- [ ] **litecoin**: See [build log](http://misc.voidlinux.de/failed/litecoin__do_build.log)
- [ ] **lmms**: See [build log](http://misc.voidlinux.de/failed/lmms__do_build.log)
- [ ] **log4cxx**: See [build log](http://misc.voidlinux.de/failed/log4cxx__do_build.log)
- [ ] **logrotate**: See [build log](http://misc.voidlinux.de/failed/logrotate__do_build.log)
- [ ] **ltris**: See [build log](http://misc.voidlinux.de/failed/ltris__do_build.log)
- [ ] **lumina**: See [build log](http://misc.voidlinux.de/failed/lumina__do_build.log)
- [ ] **mDNSResponder**: See [build log](http://misc.voidlinux.de/failed/mDNSResponder__do_build.log)
- [ ] **mac**: See [build log](http://misc.voidlinux.de/failed/mac__do_build.log)
- [ ] **milkytracker**: See [build log](http://misc.voidlinux.de/failed/milkytracker__do_build.log)
- [ ] **mixxx**: See [build log](http://misc.voidlinux.de/failed/mixxx__do_build.log)
- [ ] **most**: See [build log](http://misc.voidlinux.de/failed/most__do_build.log)
- [ ] **msp430-toolchain**: See [build log](http://misc.voidlinux.de/failed/msp430-toolchain__do_build.log)
- [ ] **mypaint**: See [build log](http://misc.voidlinux.de/failed/mypaint__do_build.log)
- [ ] **nemiver**: See [build log](http://misc.voidlinux.de/failed/nemiver__do_build.log)
- [ ] **nestopia**: See [build log](http://misc.voidlinux.de/failed/nestopia__do_build.log)
- [ ] **ngrep**: See [build log](http://misc.voidlinux.de/failed/ngrep__do_build.log)
- [ ] **nitrogen**: See [build log](http://misc.voidlinux.de/failed/nitrogen__do_build.log)
- [ ] **nomacs**: See [build log](http://misc.voidlinux.de/failed/nomacs__do_build.log)
- [ ] **nsgenbind**: See [build log](http://misc.voidlinux.de/failed/nsgenbind__do_build.log)
- [ ] **obnam**: See [build log](http://misc.voidlinux.de/failed/obnam__do_build.log)
- [ ] **octave**: See [build log](http://misc.voidlinux.de/failed/octave__do_build.log)
- [ ] **openlierox**: See [build log](http://misc.voidlinux.de/failed/openlierox__do_build.log)
- [ ] **openmsx**: See [build log](http://misc.voidlinux.de/failed/openmsx__do_build.log)
- [ ] **oprofile**: See [build log](http://misc.voidlinux.de/failed/oprofile__do_build.log)
- [ ] **phantomjs**: See [build log](http://misc.voidlinux.de/failed/phantomjs__do_build.log)
- [ ] **pipenightdreams**: See [build log](http://misc.voidlinux.de/failed/pipenightdreams__do_build.log)
- [ ] **ponyc**: See [build log](http://misc.voidlinux.de/failed/ponyc__do_build.log)
- [ ] **r2-bindings**: See [build log](http://misc.voidlinux.de/failed/r2-bindings__do_build.log)
- [ ] **ragel**: See [build log](http://misc.voidlinux.de/failed/ragel__do_build.log)
- [ ] **rawtherapee**: See [build log](http://misc.voidlinux.de/failed/rawtherapee__do_build.log)
- [ ] **rcs**: See [build log](http://misc.voidlinux.de/failed/rcs__do_build.log)
- [ ] **re2**: See [build log](http://misc.voidlinux.de/failed/re2__do_build.log)
- [ ] **sdb**: See [build log](http://misc.voidlinux.de/failed/sdb__do_build.log)
- [ ] **slrn**: See [build log](http://misc.voidlinux.de/failed/slrn__do_build.log)
- [ ] **sound-juicer**: See [build log](http://misc.voidlinux.de/failed/sound-juicer__do_build.log)
- [ ] **soundmodem**: See [build log](http://misc.voidlinux.de/failed/soundmodem__do_build.log)
- [ ] **ssr**: See [build log](http://misc.voidlinux.de/failed/ssr__do_build.log)
- [ ] **stack**: See [build log](http://misc.voidlinux.de/failed/stack__do_build.log)
- [ ] **starplot**: See [build log](http://misc.voidlinux.de/failed/starplot__do_build.log)
- [ ] **supercollider**: See [build log](http://misc.voidlinux.de/failed/supercollider__do_build.log)
- [ ] **swh-lv2**: See [build log](http://misc.voidlinux.de/failed/swh-lv2__do_build.log)
- [ ] **swift**: See [build log](http://misc.voidlinux.de/failed/swift__do_build.log)
- [ ] **teeworlds**: See [build log](http://misc.voidlinux.de/failed/teeworlds__do_build.log)
- [ ] **telepathy-haze**: See [build log](http://misc.voidlinux.de/failed/telepathy-haze__do_build.log)
- [ ] **thunderbird**: See [build log](http://misc.voidlinux.de/failed/thunderbird__do_build.log)
- [ ] **trousers**: See [build log](http://misc.voidlinux.de/failed/trousers__do_build.log)
- [ ] **ufoai**: See [build log](http://misc.voidlinux.de/failed/ufoai__do_build.log)
- [ ] **v8**: See [build log](http://misc.voidlinux.de/failed/v8__do_build.log)
- [ ] **vba-m**: See [build log](http://misc.voidlinux.de/failed/vba-m__do_build.log)
- [ ] **vdrift**: See [build log](http://misc.voidlinux.de/failed/vdrift__do_build.log)
- [ ] **virtualbox-ose**: See [build log](http://misc.voidlinux.de/failed/virtualbox-ose__do_build.log)
- [ ] **winusb**: See [build log](http://misc.voidlinux.de/failed/winusb__do_build.log)
- [ ] **wvstreams**: See [build log](http://misc.voidlinux.de/failed/wvstreams__do_build.log)
- [ ] **xautolock**: See [build log](http://misc.voidlinux.de/failed/xautolock__do_build.log)

---

<strong id="space">[1]:</strong> My web space is limited to 50GB.

**Comment by pullmoll** [comment]
Funny... trying to build pcre for i686 tells me [I need a C++ compiler](http://sprunge.us/TSjP)
I have some: either `g++` or `i686-pc-linux-gnu-c++` or even `i686-pc-linux-gnu-g++`.

**Comment by pullmoll** [comment]
Bootstrapping gcc is delicate. Bootstrapping the i686 arch is even more so.
I applied another patch for gcc from [Arch Linux](https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/gcc#n42) and now it seems that I can use a `binary-bootstrap i686` (with gcc-4.9.4) environment to build `./xbps-arch -m masterdir-x86 -N base-chroot`.

**Comment by pullmoll** [comment]
Tracked down the problem causing cross-i686-pc-linux-gnu to fail.
The error was like this
```
sysdeps/sysv/unix/linux/posix_fallocate.c:39:1: error: bp can not be used in asm here
```

It turns out we can not use `-Os` when building `glibc` because the syscall macros with 6 parameters rely on `gcc >= 5.x` optimizations to smartly replace the register `ebp` with something else. This works for `-O1`, `-O2` and `-O3`, but not for `-Os`.

**Comment by pullmoll** [comment]
### Issues: open and solved
- [x] **i686 chroot**: Built `-N base-chroot` after a `binary-bootstrap i686` from `gcc4` pkgs
- [x] **cross-i686-pc-linux-gnu**: Needed to change optimization for glibc to e.g. `-O2`
- [x] **qupzilla javascript**: Needed to add `-fno-delete-null-pointer-checks` flag to `CXXFLAGS` for `qt5`

**Comment by pullmoll** [comment]
Help wanted with the list of issues above! :construction_worker:
You could use one of the live ISOs in a VM, clone the repo and take a look at the problems, if you have time to spare...

**Comment by chneukirchen** [comment]
https://groups.google.com/forum/#!topic/golang-nuts/a4UHUIk3riw suggests an out-of-tree build help for gcc go.

**Comment by chneukirchen** [comment]
A plain bootstrap from your repo fails for me currently with (on glibc)

```
checking for suffix of object files... configure: error: in `/data/chris/tmp/void-packages-gcc6.2.1/masterdir/builddir/gcc-6.2.1/x86_64-unknown-linux-gnu/libgcc':
configure: error: cannot compute suffix of object files: cannot compile
See `config.log' for more details.
make[2]: *** [Makefile:16864: configure-stage1-target-libgcc] Error 1
make[2]: Leaving directory '/data/chris/tmp/void-packages-gcc6.2.1/masterdir/builddir/gcc-6.2.1'
make[1]: *** [Makefile:21010: stage1-bubble] Error 2
make[1]: Leaving directory '/data/chris/tmp/void-packages-gcc6.2.1/masterdir/builddir/gcc-6.2.1'
make: *** [Makefile:915: all] Error 2
```

**Comment by pullmoll** [comment]
Uh, ok, I don't have the `export ac_cv_objext=o` in the template or `common/environment/configure/autoconf_cache/x86_64-linux` file.
You could try to add that. Happens with gcc4 -> gcc6.

**Comment by pullmoll** [comment]
Thanks @chneukirchen ! I put the lines setting `LD_LIBRARY_PATH` into `do_build()` now.

Meanwhile I sifted through [this Alpine Linux patchset](http://git.alpinelinux.org/cgit/aports/tree/main/gcc) and appended the patches required to make gcj build for musl to `gcc/files/gcc-6.2.1-musl.diff`.

This means we now have `gcj` for both, glibc and musl, and could think about bootstrapping `openjdk` with gcj. If building `gcj` also works in the cross-* pkgs is to be tested.

**Comment by chneukirchen** [comment]
What's the plan for this? Fix the outstanding issues and then?

**Comment by pullmoll** [comment]
I currently suffer from a psychic disorder, again and decided to close down my efforts on this.
Anyone can take the start I made and improve or even finish it.
It's just that I first of all have to take care of myself. Sorry.

**Comment by ebfe** [comment]
Get well soon.

**Comment by Gottox** [comment]
@pullmoll: I hope, you get well soon! Thanks for all your work and get well soon!

**Comment by asergi** [comment]
@pullmoll take care!

**Comment by chneukirchen** [comment]
@pullmoll Thanks for all your efforts and I hope you recover soon.

**Comment by pullmoll** [comment]
FYI now taking a higher dose of [my pills](https://en.wikipedia.org/wiki/Quetiapine) ([German](http://de.wikipedia.org/wiki/Quetiapin)) :-/
Life sometimes isn't fun at all, while fortunately there's pharmaceutical research.

**Comment by raedwulf** [comment]
Sad to hear this @pullmoll I can have a look at this tomorrow.

**Comment by pullmoll** [comment]
Back to *normal* brain operating ;-P

Answering @chneukirchen 's question:
> What's the plan for this? Fix the outstanding issues and then?

It looks like the gcc-6.2.1 created packages can be mixed with gcc-4.9.4 created ones thanks to the 
`--with-default-libstdcxx-abi=gcc4-compatible` configure option.

Wheter we should start with a homegrown gcc-6.2.1 or wait until it becomes official is a decision on its own. We could perhaps use the latest git revision instead of the one used in Arch Linux.

Then we could *simply* switch to gcc-6.2.1 by first doing the `x86_64` bootstrap and then creating the `base-system`, `base-devel` and the `cross-*` packages in a trusted environment.

After the `cross-*` packages are built, we can also create the `base-chroot` or `base-chroot-musl` packages for the arches we support, so every arch can be `binary-bootstrap`ped.

Then `glibc` for `i686` has to be built before `gcc-multilib` for x86_64.

All the packages built could then be *injected* into the main repository by copying and doing `xbps-rindex -af ${pkgname}.xbps`. Care must be taken to properly register the `*.noarch.xbps` packages for each arch which requires them.

A probably better choice would be to create a new primary repository like `repo.voidlinux.eu/gcc6` with the current `repo.voidlinux.eu/current` becoming a fallback for packages which are not yet built using gcc6. This is less invasive and doesn't have the risk of smashing the entire repository infrastructure.

There is, however, currently one show stopper which I was not able to track down:
The `misp*-musl` arches fail to build `libffi` because the assembler thinks the target arch is `mips1` and cannot assemble some of the `*.S` files. This seems to be due to a missing (or new, extra) piece in the `gcc-6.2.1-musl.diff` file.

Also trying to build gccgo for musl libc fails. Everything compiles ok with the musl patches applied, but linking fails to resolve some symbols which seem to be contained in `libgcc.a`.

From what I've learned by building and trying out packages built with gcc-6.2.1 both, on hardware and in VMs, I conclude that switching is possible and not too cumbersome, even if some packages may need (new) patches with their next update.

**Comment by Gottox** [comment]
I think we should do a hard cut here:

1. stop the builder
2. cp -al /hostdir/binpkgs /hostdir/binpkgs.gcc4
3. ln -sf /hostdir/binpkgs.gcc4 /srv/repo.voidlinux.eu/current
4. ln -sf /hostdir/binpkgs /srv/repo.voidlinux.eu/current-gcc6
5. start the builder
6. **update gcc**
7. manually test updates
8. ln -sf /hostdir/binpkgs /srv/repo.voidlinux.eu/current
9. *profit*

**Comment by pullmoll** [comment]
I don't think bootstrapping gcc6 will work without manual intervention.
Care must be taken regarding the sequence of packages to build.

Also it may make sense to (force) re-build the bootstrap pkgs after the bootstrap is done. This means to have them built with gcc6, not gcc4. They *should* be equal but will most probably be binary different.

**Comment by Gottox** [comment]
I moved "start the builder" after "rebuild affected packages" so we can boostrap manually.

**Comment by Gottox** [comment]
force rebuilding can simply be done on "rebuild affected packages" with a revbump.

**Comment by pullmoll** [comment]
Ok, a revbump of the bootstrap packages seems to make sense.
Here's a list of steps required - not tested, just dumped from my brain :-)
```
# Merge the head commit of my fork
./xbps-src zap
./xbps-src bootstrap
# Revbump pkgs listed in base-chroot and base-chroot itself
./xbps-src -N pkg base-chroot
# Now base-chroot is built with gcc6
./xbps-src zap
./xbps-src binary-bootstrap
./xbps-src -N pkg base-system
./xbps-src -N pkg base-devel
find srcpkgs -name "cross-*" -type d -printf "%f\n" | xargs -t -n1 ./xbps-src -N pkg
for arch in armv5tel armv6l armv7l aarch64 i686; do
    ./xbps-src -a ${arch} -N base-chroot
done
./xbps-src -N gcc-multilib
for arch in armv6l armv7l aarch64 i686 mips mipsel mipselhf; do
    ./xbps-src -a ${arch}-musl -N base-chroot-musl
done
# Now things should work without intervention
```
I don't know for sure if the `cross-*` pkgs also need to be built with `i686` and `x86_64-musl`...

**Comment by Gottox** [comment]
I'm currently doing an fbulk build and hope to get an updated list with failing packages.

**Comment by raedwulf** [comment]
This may be relevant about the MIPS assembling issue:
https://gcc.gnu.org/ml/gcc-patches/2016-08/msg01757.html

**Comment by pullmoll** [comment]
@raedwulf The fact that more patches for mips*-musl make it into gcc makes me hope that a newer gitrev will just work without us having to apply more patches.

**Comment by Gottox** [comment]
I currently run another fbulk build on this branch.

**Comment by pullmoll** [comment]
I have built 6754 (x86_64), 6655 (i686) and 6475 (x86_64-musl) packages with it.
There are only some dozen packages left which need (sometimes trivial) fixing.

**Comment by Gottox** [comment]
Nice to hear. most if the errors I encountered yet are 404s on the distfiles.

**Comment by pullmoll** [comment]
BTW perhaps someone with access to the official http://sources.voidlinux.eu could create another `gcc-6.2.1.tar.xz` based on (one of)  the latest git revision(s)...

**Comment by raedwulf** [comment]
Can you update the list of packages that need patching to work with gcc 6.2.1?

**Comment by pullmoll** [comment]
I can post update to a partial list of what has been tried and failed.

**Comment by raedwulf** [comment]
Thanks :) That would be great, I patched audacity already but xbps-fbulk is a bit inconvenient for me atm to build my own list.  Should I just throw in pull requests here for gcc-6.2.1 patches?  As far as I know, they should work with gcc-4.9 as well, normally it's just the compiler being more picky and bad code in the relevant packages.

**Comment by pullmoll** [comment]
I don't use xbps-fbulk but my own script to list all pkgs and build them for the three arches, using `-E -N` parameters to skip what's already done and to use the local repository only.

Sure, you can create PRs for fixes for the gcc6 build. You should verify the changes do either not change the gcc4 build (gcc6 only C{,XX}FLAGS) or revbump the pkg and build for both compilers.

**Comment by pullmoll** [comment]
Now x86_64 is through once with 7060 of 7384 packages built. There are quite a few distfile downloads failing and there's the list of failures at the top.

**Comment by pullmoll** [comment]
Failing downloads:
- [ ] apache-directory-studio-2.0.0.v20130628_3: http://wwwftp.ciril.fr/pub/apache/directory/studio/dist/2.0.0.v20130628/ApacheDirectoryStudio-linux-x86_64-2.0.0.v20130628.tar.gz: Not Found

**Comment by Gottox** [comment]
fixed a few URLs.

**Comment by pullmoll** [comment]
Updating gcc6 to b8c3476
```
distfiles="http://distfiles.voidlinux.de/gcc-6.2.1/gcc-6.2.1.tar.xz"
checksum=1379359546ef61a08bcf56d423a76900de3202d129c84ccb347ff39f57badbf5
```
And commencing another bootstrap to see if anything changes.

**Comment by Gottox** [comment]
All checksums are updated, chrome-widevine can be ignored.

@pullmoll: What are the next steps to take?

**Comment by pullmoll** [comment]
As I outlined above someone with access to the https://sources.voidlinux.eu/ server should create a semi-official gcc-2.6.1.tar.xz package. I would then change the forked repository again to use that distfile and create a PR.

The switch then can be rather smooth. I think just x86_64 base-chroot hast to be built manually, twice with the revbumps, and i686 base-chroot should be built from a gcc4 bootstrapped masterdir.

Then the revbumps on the cross-* packages should do their work and the bots can continue to do their job.

**Comment by Gottox** [comment]
gcc shouldn't go to sources but to repo.voidlinux.eu/dist.

Is there any schedule when an official release of 2.6.1 is planned?

**Comment by pullmoll** [comment]
I haven't seen one. If it happens, I think we can just revbump gcc then.

**Comment by pullmoll** [comment]
Edited the above list of things to do *offline*. The `cross-*` pkgs don't have to be built manually I think.

**Comment by pullmoll** [comment]
Just let me know when https://repo.voidlinux.eu/distfiles/gcc-6.2.1.tar.xz is there and I'll create the PR for gcc6.

**Comment by Gottox** [comment]
```
git clone git://gcc.gnu.org/git/gcc.git gcc-6.2.1
cd gcc-6.2.1
git checkout 6f2035ca9fb6fbdd0686a926e79caad44b3b49f1
cd ..
find gcc-6.2.1 \! -type d -print0 | \
    sort -z | \
    grep -z -v \.git/ | \
    tar cvf - -T - | \
    xz -v -9 -T 0 > gcc-6.2.1.tar.xz
```

**Comment by Gottox** [comment]
```
distfiles="https://repo.voidlinux.eu/distfiles/gcc-6.2.1.tar.xz"
```

**Comment by pullmoll** [comment]
checksum=a6fb95d4afb88f2cb6fbf6263b8f833d9c75a66fe89162f8f3cadf0cc49a8404

Modifying the files and creating a PR now.

**Comment by pullmoll** [comment]
TODO after the wrong changes because this distfile was actually gcc-7.0.0:
Restore the patch `libcpp-source_date_epoch.patch` in the following files:
+ srcpkgs/cross-aarch64-linux-gnu/template
+ srcpkgs/cross-aarch64-linux-musl/template
+ srcpkgs/cross-arm-linux-gnueabi/template
+ srcpkgs/cross-arm-linux-gnueabihf/template
+ srcpkgs/cross-arm-linux-musleabi/template
+ srcpkgs/cross-arm-linux-musleabihf/template
+ srcpkgs/cross-armv7l-linux-gnueabihf/template
+ srcpkgs/cross-armv7l-linux-musleabihf/template
+ srcpkgs/cross-i686-linux-musl/template
+ srcpkgs/cross-i686-pc-linux-gnu/template
+ srcpkgs/cross-mips-linux-musl/template
+ srcpkgs/cross-mipsel-linux-musl/template
+ srcpkgs/cross-mipsel-linux-muslhf/template
+ srcpkgs/cross-x86_64-linux-musl/template

Create symlinks to `srcpkgs/gcc/patches/libcpp-source_date_epoch.patch` in above pkgs `./files` directories.

**Comment by Gottox** [comment]
sorry for this. Updated the distfile, new shasum: 3f3e1352e785622f57c7cdbdc978de33f3f9b97c3085ba85c6486a23fd8876ba

**Comment by pullmoll** [comment]
@Gottox You need to `git checkout gcc-6-branch`. I forgot to mention this in my first post.

**Comment by Gottox** [comment]
yep, I fixed it and created a file based on 27ad3c7127ce4e40147fbf136435cca9c7963dbb

**Comment by pullmoll** [comment]
Ok, I'll change that, restore the patches and create a PR again, once I verified that at least bootstrapping works. This may take longer than tonight and I intend to go to a concert later that afternoon :)

**Comment by Gottox** [comment]
If it's about restoring and you're using git, git-reflog + git-reset --hard may help. Anyway, hf at the concert!

**Comment by pullmoll** [comment]
The concert is this one http://www.zone-bonn.de/aktuelles.html

I think that while *offline* also the `base-chroot-musl` needs a build with its own masterdir.
```
./xbps-src -m musl zap
./xbps-src -m musl binary-bootstrap x86_64-musl
./xbps-src -m musl -N pkg base-chroot-musl
```
All this before revbumping the base-chroot and base-chroot-musl pkgs.

**Comment by pullmoll** [comment]
Bad news: the package `bison` fails to build here with a `perl` compiled with `gcc6` in the first round. Building `perl` stopped with an error and I just called `./xbps-src -N pkg base-chroot` again.
Only after rebuilding `perl` with `./xbps-src -f pkg perl` the 2nd build of `base-chroot` packages continues.

Perhaps `perl` needs `disable_parallel_builds=yes`?

**Comment by pullmoll** [comment]
Updated the steps to do offline to what I think should be necessary.

**Comment by biax** [comment]
@Gottox  "Is there any schedule when an official release of 6.2.1 is planned?"

Yeah I and a lot of people are waiting for gcc 6.2x and then we can migrate everything from arch to void :)

**Comment by chneukirchen** [comment]
FTR, there will be no official 6.2.1 release, the next gcc release is 6.3.

**Comment by pullmoll** [comment]
With the latest gcc-6.2.1 some pkgs fail to cross compile (for aarch64):
- [ ] webkit2gtk
- [ ] kwindowsystem

**Comment by pullmoll** [comment]
Another major dropout is boost. For example libabw fails to link against boost for missing symbols:
```
  CXXLD    abw2raw
../../lib/.libs/libabw-0.1.so: undefined reference to `boost::system::system_category()'
../../lib/.libs/libabw-0.1.so: undefined reference to `boost::system::generic_category()'
```

I have no solution yet. Trying to build boost with `CXXFLAGS="-std=c++14"` (seen in Arch Linux) does not solve the problem.

**Comment by chneukirchen** [comment]
Was the libabw thing related to https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=835280 ?

**Comment by pullmoll** [comment]
@chneukirchen Yes, it was the same error that occured here with gcc6. The patch adds the required flags and works for both, gcc4 and gcc6.

AFAICT there are no showstoppers anymore.

**Comment by pullmoll** [comment]
That was too quick. None of the x86_64-musl live images with a DE are working right in `qemu`. They all display a black screen sooner or later, with the *best result* coming from the `enlightenment` and `lxqt` ISOs. There the black screen shows after you start an application like `firefox` or `qupzilla`. A terminal in `enlightenment` does work, though.

Unless this is related to `qemu` itself, we have a serious problem with some part of `Xorg` or its libraries and the Musl libc with gcc-6.2.1 :unamused:

**Comment by biax** [comment]
while you're at it, are you able to check if mtr-musl works? a week or so ago I remember coming to conclude that mtr-musl always gives: no domainname or something, while fine with mtr glibc

**Comment by aeosynth** [comment]
do we have to change a repo setting to get the gcc 6 packages, or does this iso pre-set it for ur?

**Comment by pullmoll** [comment]
@biax I get an error "No nameservers defined." and then mtr exits.

@aeosynth You have to create e.g. `/etc/xbps.d/00-gcc6.conf` with `repository=http://repo.voidlinux.de/current` or `repository=http://voidlinux.myds.me/repo/current` to receive the gcc6 compiled packages. Remember this is meant for testing, not for production use.

**Comment by biax** [comment]
@pullmoll exactly. that doesnt happen in the glibc version I believe.

**Comment by pullmoll** [comment]
@biax This is an unrelated topic, though, so please let's not discuss here.

FWIW The error I see in `/var/log/messages` when running `void-live-x86_64-musl-20161108-kde.iso` in `qemu` and logging in to KDE with `anon` and `voidlinux` is this:
```
... kdm[767]: X server for display :0 terminated unexpectedly
```

**Comment by pullmoll** [comment]
FYI I'm restarting once more with today's gcc-6.2.1 to see if recent changes fix xorg for musl.
```
distfile=http://distfiles.voidlinux.de/gcc-6.2.1/gcc-6.2.1.tar.xz
checksum=6bc5af0d2138bb8492c62ced00776b76bbc98b7297287aaa55adbe7d07786ac8
```

**Comment by ebfe** [comment]
@pullmoll 
```
void-live-x86_64-20161108-enlightenment.iso
void-live-x86_64-20161108-xfce.iso
```
same problems with qemu, seems to work fine in virtualbox.

**Comment by ebfe** [comment]
```void-live-x86_64-musl-20161108-kde.iso``` is throwing me back to the login screen after a few seconds. Since this is different, maybe this is simply something crashing due to musl stack sizes?

**Comment by pullmoll** [comment]
@ebfe I think the KDE image does not have enough free space to accomodate for the files created during the first start. I already patched my local copy of `mklive.sh.in` to add some more slack space to the file system, so I'll add even more for the next round.

Good to  know that the images are working in virtualbox! I'll try that once all required packages to create new ISO images are built.

**Comment by pullmoll** [comment]
@ebfe For `void-live-x86_64-musl-20161113-lxqt.iso` boots in virtualbox, but as soon as I start e.g. `qterminal` the screen turns black and I can't see anything, not even if I switch to a console with `<R-ctrl>+<F2>`.

Can you verify again?

**Comment by pullmoll** [comment]
Ok, now tested on real hardware and there the lxqt image for x86_64-musl works just fine.
So it seems there is no issue with gcc6 at all, but rather `qemu` and/or `virtualbox`.

**Comment by chneukirchen** [comment]
Shall we merge it this weekend?

**Comment by pullmoll** [comment]
Yes, it is good enough to be merged, if you ask me.

**Comment by chneukirchen** [comment]
@pullmoll want to help us merge today?

**Comment by pullmoll** [comment]
TODO: create another official `gcc-6.2.1.tar.xz`
```
sed -i srcpkgs/*/template -e's;6bc5af0d2138bb8492c62ced00776b76bbc98b7297287aaa55adbe7d07786ac8;<new checksum>;'
sed -i srcpkgs/*/template -e's;http://distfiles.voidlinux.de/gcc-$version/gcc-$version.tar.xz;<new download location>;'
sed -i srcpkgs/*/template -e's;http://distfiles.voidlinux.de/gcc-${_gcc_version}/gcc-${_gcc_version}.tar.xz;<new download location>;'
```

**Comment by pullmoll** [comment]
Finished. Existing issues with (some) packages will have their own issues entry and/or PR.

**Comment by Gottox** [comment]
Thanks a lot @pullmoll. Without you this couldn't be done!

**Comment by pullmoll** [comment]
No problem. The main reason I did this was to avoid having to hack around MAME. Version 0.180 coming soon and would probably no longer compile with gcc4 ;-)
